contracts/MasterChef.sol

```diff
@@ -0,0 +1,294 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.7.0;
+
+// import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+// import "@openzeppelin/contracts/math/SafeMath.sol";
+// import "./library/TransferHelper.sol";
+// import "@openzeppelin/contracts/access/Ownable.sol";
+
+import './mocks/CryptionNetworkToken.sol';
+import './lib/EventProof.sol';
+import './lib/RLPReader.sol';
+
+contract Farm01 is Ownable{
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+    using RLPReader for bytes;
+    using RLPReader for RLPReader.RLPItem;
+
+    /// @notice information stuct on each user than stakes LP tokens.
+    struct UserInfo {
+        uint256 amount;     // How many LP tokens the user has provided.
+        uint256 rewardDebt; // Reward debt.
+    }
+    
+    /// @notice all the settings for this farm in one struct
+    struct FarmInfo {
+        IERC20 lpToken;
+        IERC20 rewardToken;
+        uint256 startBlock;
+        uint256 blockReward;
+        uint256 bonusEndBlock;
+        uint256 bonus;
+        uint256 endBlock;
+        uint256 lastRewardBlock;  // Last block number that reward distribution occurs.
+        uint256 accRewardPerShare; // Accumulated Rewards per share, times 1e12
+        uint256 farmableSupply; // set in init, total amount of tokens farmable
+        uint256 numFarmers;
+    }
+    
+    /// @notice farm type id. Useful for back-end systems to know how to read the contract (ABI) 
+    /// as we plan to launch multiple farm types
+    uint256 public farmType = 1;
+   
+    address public farmGenerator;
+
+    FarmInfo public farmInfo;
+    
+    /// @notice information on each user than stakes LP tokens
+    mapping (address => UserInfo) public userInfo;
+
+    // Receipts root vs bool. It is used to avoid same tx submission twice.
+    mapping (bytes32 => bool) public receipts;
+
+    uint256 public constant EVENT_INDEX_IN_RECEIPT = 3; 
+    uint256 public TRANSFER_EVENT_PARAMS_INDEX_IN_RECEIPT = 1;
+
+    bytes32 public constant TRANSFER_EVENT_SIG = 0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef;
+
+    address CRYPTION_VALIDATOR_SHARE_PROXY_CONTRACT = address(0x44d2c14F79EF400D6AF53822a054945704BF1FeA);
+
+    event Deposit(address indexed user, uint256 amount);
+    event Withdraw(address indexed user, uint256 amount);
+    event EmergencyWithdraw(address indexed user, uint256 amount);
+
+    constructor( address _farmGenerator) public {
+        farmGenerator = _farmGenerator;
+    }
+
+    /**
+     * @notice initialize the farming contract. 
+     * This is called only once upon farm creation and the FarmGenerator ensures the farm has the correct paramaters
+     */
+    function init (
```

Comment:
init method should be removed and initializations can be done in constructor.

Reply:
Done